### DISCUSSION TITLE
PeakRDL-regblock/SystemRDL Devel Guide

### DISCUSSION THREAD

**Discussion by galaviel** [SystemRDL/systemrdl-compiler#165] [General]
Hi,

I want to play around with the code of those 2 projects using Eclipse/Pydev.

Is there a simple guide on how/what I need to setup?

I've cloned both PeakRDL and SystemRDL and PeakRDL-regblock from github.

any help appreciated,

thanks!

**Comment by amykyta3** [comment]
I don't really have a formal developers guide for these projects. However I do occasionally put my personal notes into the repository. Perhaps these notes will help you get oriented.

For the `systemrdl-compiler` project, I have the notes published here: https://systemrdl-compiler.readthedocs.io/en/stable/dev_notes/logbook.html

For `PeakRDL-regblock`, the notes are in the repository, just less organized at the moment: https://github.com/SystemRDL/PeakRDL-regblock/tree/main/docs/dev_notes

The command line wrapper `PeakRDL` is generally less complicated, so I do not have any dev notes for that one. I do plan on writing a proper docs page that describes the high-level relationship between `PeakRDL`, and the various `PeakRDL-*` plugins.

**Comment by amykyta3** [comment]
Regarding setup while developing, there really isn't much. 
I usually create a Python virtual environment, and then install the components in editable mode (pip install -e path/to/thing).

For most projects, there is a suite of unit-tests you can run under the tests directory.
Regblock depends on some EDA tools to be installed for the tests to work. More details on that here:  https://github.com/SystemRDL/PeakRDL-regblock/tree/main/tests#readme

**Comment by galaviel** [comment]
Thanks allot for the link ! great I'll look..

Curious though how you commit/push the files back to the main repo?
do you edit the Python files inside the virtual env, then copy them back to the clone, commit and push ?

**Comment by galaviel** [comment]
I couldn't run / debug 'PeakRDL\src\peakrdl\main.py' directly because I got tangled up with relative imports:

```
Traceback (most recent call last):
  File "C:\py\PeakRDL\src\peakrdl\main.py", line 9, in <module>
    from .__about__ import __version__
ImportError: attempted relative import with no known parent package
```
so I ended up creating a small, external script (pasted below). Not sure if this is the recommended way:

```
from peakrdl import main as mn
import sys

def main():
    print(sys.argv)
    sys.argv.extend(["regblock", "CMDQ_CSR.csv.rdl", "-o", "regblock_cmdq", "--cpuif", "apb3-flat"])
    mn.main()

if __name__ == '__main__':
    main()
```

**Reply by amykyta3** [reply]
Odd that you had to resort to a wrapper script like that. Usually when I'm developing, I'll install in "editable" mode (using `-e`) as follows:
```
python3 -m pip install -e path/to/peakrdl
```
If developing across multiple PeakRDL projects, this works well if you install them in dependency-order.

This allows you to run the command-line tool directly off of the cloned repository.

**Comment by galaviel** [comment]
Perhaps because of Eclipse... what are you using to debug?
When I'm done with committing everything I'll try again to re-verify issue still exists.
My recepie (Python3.8 on Windows and Eclipse/PyDev):
```

# create venv
python -m venv venv1

# active the virtual env
venv1\Scripts\activate

# add the python.exe under C:\py\venv1\Scripts\python.exe as the Pydev selected interperter

# clone
# in order to install with -e (edit) you must clone locally all those repo's (I think..)
# perhaps the order here is wrong??
git clone https://github.com/SystemRDL/PeakRDL
git clone https://github.com/SystemRDL/PeakRDL-regblock
git clone https://github.com/SystemRDL/systemrdl-compiler

# install from the cloned location
pip install -e PeakRDL
pip install -e PeakRDL-regblock
pip install -e systemrdl-compiler

```

**Reply by amykyta3** [reply]
Looks ok, but yeah you want to reverse the order of your pip installs, otherwise pip will fetch the published packages from pypi

Try this:
```
pip install -e systemrdl-compiler
pip install -e PeakRDL-regblock
pip install -e PeakRDL
```

then pip will see that the dependency is already satisfied by the editable install, and will not fetch the official one.
